### PR TITLE
Add password generator action

### PR DESCRIPTION
**Pull request by TonioRyo**
Add an action to generate securely a password during a playbook run to use with Entra ID and Active Directory password reset actions

### CODE REVIEW

**Review by squioc — CHANGES_REQUESTED** [review]
Thanks for the action.
Can you move this action to the automation module [utils](https://github.com/SEKOIA-IO/automation-library/tree/main/Utils), that is more suitable for this kind of actions?

File: Utils/utils/password_generator.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import secrets
+import string
+from sekoia_automation.action import Action
+
+
+class PasswordGenerator(Action):
+    def run(self, arguments: dict):
+        password_length = self.json_argument("password_length", arguments)
+        number_of_digits = self.json_argument("number_of_digits", arguments)
+        number_of_special_characters = self.json_argument("number_of_special_characters", arguments)
```
Comment (Darkheir) [inline Utils/utils/password_generator.py:10]: No need to fetch it from a file also setting defaults would be nice
```suggestion
        password_length = arguments.get("password_length", 20)
        number_of_digits = arguments.get("number_of_digits", 1)
        number_of_special_characters = arguments.get("number_of_special_characters", 1)
```

Comment (Darkheir) [inline Utils/utils/password_generator.py:10]: We also need to make sure that `number_of_digits` + `number_of_special_characters` < `password_length`

File: Utils/utils/password_generator.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        while True:
+            password = ""
+            for i in range(password_length):
+                password += "".join(secrets.choice(alphabet))
+
+            if (
+                sum(char in special_chars for char in password) >= number_of_special_characters
+                and sum(char in digits for char in password) >= number_of_digits
+            ):
+                break
```
Comment (Darkheir) [inline Utils/utils/password_generator.py:27]: This method is not efficient, especially when `number_of_special_characters` or `number_of_digits` are getting close to `password_length`
```suggestion
        chars = [secrets.choice(special_chars) for _ in range(number_of_special_characters)]  # n special chars
        chars += [secrets.choice(digits) for _ in range(number_of_digits)] # n digits
        chars += [secrets.choice(alphabet) for _ in range(password_length - num_special - num_digits)]  # fill remaining with random
        random.shuffle(chars)  # Shuffle it so char and digits are not always at the beginning
        password = "".join(chars)
```

**Review by Darkheir — CHANGES_REQUESTED** [review]

File: Utils/action_password_generator.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                "description": "Number of special characters in the password",
+                "default": 1,
+                "type": "integer"
+            }
+        },
+        "required": [
+            "password_length",
+            "number_of_digits",
+            "number_of_special_characters"
+        ]
```
Comment (Darkheir) [inline Utils/action_password_generator.json:31]: Not required anymore :-)

```suggestion
```

File: Utils/utils/password_generator.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class PasswordGenerator(Action):
+    def run(self, arguments: dict):
+        password_length = arguments.get("password_length", 20)
+        number_of_digits = arguments.get("number_of_digits", 1)
+        number_of_special_characters = arguments.get("number_of_special_characters", 1)
+
+        if number_of_digits + number_of_special_characters > password_length:
+            raise ValueError("number_of_digits + number_of_special_characters must be lower than password_length")
```
Comment (Darkheir) [inline Utils/utils/password_generator.py:14]: Raising an exception will result in having the errors message `An unexpected error occured: ...`.

```suggestion
            self.error("number_of_digits + number_of_special_characters must be lower than password_length")
            return
```

File: Utils/tests/test_password_generator.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    assert sum(char in digits for char in results.get("password")) >= 5
+
+
+def test_password_generator_failure(arguments_failure):
+    action = PasswordGenerator()
+
+    try:
+        results = action.run(arguments_failure)
+    except ValueError as e:
+        assert "number_of_digits + number_of_special_characters must be lower than password_length" in str(e)
```
Comment (Darkheir) [inline Utils/tests/test_password_generator.py:36]:
```suggestion
    action.run(arguments_failure)
    assert action._error == "number_of_digits + number_of_special_characters must be lower than password_length"
```

**Review by Darkheir — CHANGES_REQUESTED** [review]
Minor changes but code logic sounds good to me :-)

**Review by Darkheir — APPROVED** [review]

**Review by squioc — APPROVED** [review]

**Pull request merged**